Feat: Improve Local Firecrawl Lifecycle, Source Handling & Logging

KunihiroS opened the pull request:
This PR introduces several enhancements to the DeepResearch MCP server, focusing on improving the handling of local Firecrawl instances, managing source metadata output, refining logging, and updating documentation.

**Key Changes:**

1.  **On-Demand Local Firecrawl Management:**
    *   Implemented automatic startup (`docker-compose up -d`) and shutdown (`docker-compose down -v`) of local Firecrawl containers when the `deep-research` tool is invoked using a local `FIRECRAWL_BASE_URL`. This eliminates the need to keep containers running constantly, saving system resources.
    *   Added a health check after startup (`startLocalFirecrawl`) that polls the Firecrawl API endpoint until it's ready, replacing the previous fixed delay. This significantly reduces unnecessary waiting time.
    *   Ensured Docker volumes are removed upon shutdown (`docker-compose down -v`) to prevent disk space accumulation over multiple runs.

2.  **Separate Source Metadata Output:**
    *   Modified `writeFinalReport` to exclude the `## Sources` section from the main Markdown report.
    *   Added functionality to save the collected `sourceMetadata` array into a separate JSON file within a newly created `report_source/` directory in the project root.
    *   The path to this generated JSON file is now included in the tool's output metadata under the `sourceFilePath` key.

3.  **Improved Filename Sanitization:**
    *   Updated the filename generation logic for the source metadata JSON file. It now uses a combination of a timestamp and a SHA-1 hash of the input query (`sources_[timestamp]_[queryHash].json`) to ensure filesystem-safe names, even with non-ASCII characters in the query.

4.  **Refined Logging:**
    *   Added more granular timing logs for critical steps like Docker startup/shutdown, health checks, `deepResearch` function execution, `writeFinalReport` execution, Firecrawl searches, and `processSerpResult`.
    *   Reduced log verbosity by removing some detailed internal logs (e.g., full stdout/stderr from Docker unless an error is suspected, detailed metadata logging within functions) while retaining essential information for monitoring and debugging. Error logs now prioritize the error message.

5.  **Parameter Schema & Documentation Update:**
    *   Restored the maximum allowed values for `depth` and `breadth` parameters to 5 in the Zod schema.
    *   Added warning notes to the `depth` and `breadth` parameter descriptions, advising users that values above 3 can significantly increase execution time and potentially lead to timeouts (e.g., >10 minutes), especially when using local Firecrawl.
    *   Updated the README.md to reflect the new Docker lifecycle management feature and added an "Execution Time Notes" section detailing performance expectations and the impact of `depth`/`breadth` parameters.

These changes aim to make the server more resource-efficient, robust (especially regarding filenames), provide better performance insights through logging, and offer clearer guidance to users about execution time expectations.

KunihiroS committed:
Add initial log file for tracking deep research MCP activities and update package dependencies

- Created `deep-research-mcp.log` for logging events and errors during deep research processes.
- Updated package dependencies to version `0.25.2` from `0.23.1` in `package-lock.json`.
- Enhanced logging functionality in `deep-research.ts` & `mcp-server.ts` for better error handling and progress tracking.
- Established dummy logging for CLI context in `run.ts` for consistency.

KunihiroS committed:
Refactor source metadata handling: remove inline source generation, pass metadata to MCP server. Update logs accordingly.

KunihiroS committed:
Add detailed logging and error handling for deep research process; save source metadata to JSON.

KunihiroS committed:
Update log for Deep Research MCP with service start, research execution, and source metadata generation

- Added detailed logs for Firecrawl service operations.
- Implemented logging of queries and results for deep research.
- Created a new JSON file to store source metadata from the research.
- Adjusted logging statements for clarity and relevance.

KunihiroS committed:
Enhance Firecrawl readiness checks and add source metadata logging for quantum computing research output.

KunihiroS committed:
Update logs, adjust research depth/breadth limits, and modify Firecrawl shutdown command

- Added new log entries for the deep research process in `deep-research-mcp.log`
- Set maximum research depth and breadth to 2 in `mcp-server.ts`
- Updated Docker command to include `-v` flag for volume removal during Firecrawl shutdown

KunihiroS committed:
Update README.md with enhanced explanations for features, clarify execution time expectations, and add detailed resource efficiency notes.

Add logs for source metadata generation in the deep research process, replacing some verbose logs with concise entries.

Expand the breadth and depth parameter limits in the MCP server tool, updating restrictions based on performance considerations.

Adjust logging for errors to focus on concise messaging while maintaining critical information for troubleshooting.

shemhamforash23 commented on threads-mcp-server.log:
log files should be in `.gitignore`

shemhamforash23 commented on deep-research-mcp.log:
log files should be in `.gitignore`

shemhamforash23 commented on report_source/sources_2025-04-05T08-25-28-298Z___________________.json:
`report_source` directory would contain unique files for each user. 
Hence it should be in `.gitignore`.

shemhamforash23 commented on src/deep-research.ts:
```diff
@@ -81,12 +89,14 @@ async function generateSerpQueries({
   learnings,
   learningReliabilities,
   researchDirections = [],
+  log, // Add log parameter
```
unnecessary comment. it does not carry any meaningful load.

shemhamforash23 commented on src/deep-research.ts:
```diff
@@ -81,12 +89,14 @@ async function generateSerpQueries({
   learnings?: string[];
   learningReliabilities?: number[];
   researchDirections?: ResearchDirection[];
+  log: (...args: any[]) => Promise<void>; // Add log type
```
unnecessary comment. it does not carry any meaningful load.

shemhamforash23 commented on src/deep-research.ts:
```diff
@@ -1,3 +1,11 @@
+import fs from 'fs/promises';
+import path, { dirname } from 'path';
+// import { fileURLToPath } from 'url'; // Already imported below
```
unused imports should be removed

shemhamforash23 commented on src/mcp-server.ts:
```diff
@@ -1,19 +1,171 @@
+    .map(arg => (typeof arg === 'string' ? arg : JSON.stringify(arg)))
+    .join(' ');
+  try {
+    // Ensure the log directory exists (optional, fs.appendFile creates the file but not dirs)
```
duplicate comment

shemhamforash23 commented on src/mcp-server.ts:
```diff
@@ -1,19 +1,171 @@
+
+// --- Docker Control Functions ---
+
+// Docker 起動関数
```
non-English comment

shemhamforash23 commented on src/mcp-server.ts:
```diff
@@ -1,19 +1,171 @@
+  }
+}
+
+// Docker 停止関数
```
non-English comment

shemhamforash23 commented on src/mcp-server.ts:
```diff
@@ -48,52 +215,95 @@ server.tool(
+      const reportGenEndTime = Date.now();
+      await log(`[MCP Server] writeFinalReport completed. Duration: ${reportGenEndTime - reportGenStartTime}ms`);
+
+      // ▼▼▼ デバッグログ (Keep for now) ▼▼▼
```
non-English comment

shemhamforash23 commented on src/mcp-server.ts:
```diff
@@ -111,14 +332,14 @@ async function main() {
     const transport = new StdioServerTransport();
     await server.connect(transport);
-    log('Deep Research MCP Server running on stdio');
+    await log('Deep Research MCP Server running on stdio'); // await を追加
```
non-English comment

shemhamforash23 commented on src/deep-research.ts:
```diff
@@ -218,13 +226,15 @@ async function processSerpResult({
   numFollowUpQuestions = 3,
   reliabilityThreshold = 0.3,
   researchGoal = '',
+  logToFile, // Rename parameter to logToFile
```
the variable is already renamed. unnecessary comment

shemhamforash23 commented on src/deep-research.ts:
The file contains few unnecessary comments, which are not carry any meaningful load.
Please consider removing redundant comments and maybe add js-docs for functions if you want to explain why it has it's arguments.

shemhamforash23 commented on src/mcp-server.ts:
This file also needs some reworking of the comments.

shemhamforash23 reviewed (commented):
Auto docker compose start is a very useful feature to have.
The code looks good, it just needs a few finishing touches.
Keep up the good work!

KunihiroS committed:
Update .gitignore and refactor logging parameters in TypeScript files

- Add log files and report source directory to .gitignore
- Clean up logging parameter and type definitions in deep-research.ts
- Restore log statement in mcp-server.ts and remove commented debug logs

KunihiroS commented:
@shemhamforash23 Thank you for your review , and sorry for my carelessness.

I have addressed the following points you raised:

*   Updated `.gitignore` to exclude log files and the `report_source/` directory.
*   Removed unnecessary comments from `src/deep-research.ts` and `src/mcp-server.ts`.
*   Resolved the TypeScript errors by running `npm install`.

Could you please review the changes again? I appreciate that.

shemhamforash23 commented on .gitignore:
```diff
@@ -42,4 +42,11 @@ yarn-error.log*
+tasks.md
+
+# Log files
+deep-research-mcp.log
```
It's better to use wildcard here: `*.log`
This will ensure that any file with `.log` extension will be ignored in future.

shemhamforash23 commented:
@KunihiroS your changes still contain a lot of unrelated files:
- `deep-research-mcp.log`
- `report_source/sources_2025-04-05T08-25-28-298Z___________________.json`
- `report_source/sources_2025-04-05T08-42-55-645Z____AI______.json `
- `report_source/sources_2025-04-05T08-57-24-380Z_5ece637247.json` 
- `report_source/sources_2025-04-05T09-21-57-020Z_e9a0992acf.json`
- `report_source/sources_2025-04-05T09-48-15-990Z_19917a7729.json`
- `report_source/sources_2025-04-05T10-13-55-342Z_7d908e45d2.json`

Think of it like this: are these file changes necessary to be merged into main branch and required for an application to work? 
Or these are your personal content and is making sense only for you?

If it's latter, then it should be removed from PR.

KunihiroS committed:
chore: remove personal log and report files from tracked files and update .gitignore

KunihiroS committed:
chore: remove threads-mcp-server.log from tracked files

KunihiroS commented:
@shemhamforash23 
Thank you very much for your review and for pointing out the issues — I really appreciate your attention to detail.

"your changes still contain a lot of unrelated files:

deep-research-mcp.log

report_source/....json

...

If it's latter, then it should be removed from PR."

You're absolutely right — I included several personal and intermediate files that should never have been part of the PR. I sincerely apologize for the oversight.

I've now removed all of the following from the tracked files using git rm --cached:

deep-research-mcp.log
All .json files under report_source/
"It's better to use wildcard here: *.log"

Thank you for this suggestion as well. I've updated .gitignore to use *.log so that all future log files will be properly ignored.

I’ve also triple-checked!! the repository to ensure that:

No unnecessary files are being tracked
.gitignore is correctly configured
The PR now only includes the intended changes
I truly appreciate your patience and guidance. Now it is completely ready to merged, I am sure of it.
